From b430c70d6dc27931ed2c6827dce9f52204f1bee0 Mon Sep 17 00:00:00 2001
From: Joerg Roedel <jroedel@suse.de>
Date: Tue, 19 Aug 2014 17:52:03 +0200
Subject: [PATCH] PM / hibernate: Use put_page to free ring-buffer pages
References: bnc#892004, bnc#890373
Patch-mainline: Never, fix for SLES-only PG_waiters patch

Use put_page to free ring-buffer pages in load_image_lzo.
This makes sure that any stale PG_waiters bit is cleared
before the page is returned to the page allocator.

Signed-off-by: Joerg Roedel <jroedel@suse.de>
---
 kernel/power/swap.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/kernel/power/swap.c
+++ b/kernel/power/swap.c
@@ -1361,8 +1361,10 @@ out_finish:
 	}
 	swsusp_show_speed(&start, &stop, nr_to_read, "Read");
 out_clean:
+	/* Use put_page here to make sure any stale page flags get cleared */
 	for (i = 0; i < ring_size; i++)
-		free_page((unsigned long)page[i]);
+		put_page(virt_to_page(page[i]));
+
 	if (crc) {
 		if (crc->thr)
 			kthread_stop(crc->thr);
